### PR TITLE
Improve feature mode switch process

### PR DESCRIPTION
**Pull request by lixiaoyuner**
Signed-off-by: Yun Li <yunli1@microsoft.com>

#### Why I did it
Improve feature mode switch process
#### How I did it
Refine ctrmgrd service, feature service and start service inside feature container
#### How to verify it
Connect to k8s cluster as worker and scheduled by k8s 
#### Which release branch to backport (provide reason below if selected)
Fix existing bugs

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [x] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205

#### Description for the changelog

Refine ctrmgrd service, feature service and start service inside feature container

### CODE REVIEW

File: src/sonic-ctrmgrd/ctrmgr/ctrmgrd.py [hunk 1]
```diff
@@ -88,18 +88,20 @@
...
     JOIN_LATENCY: 10,
     JOIN_RETRY: 10,
     LABEL_RETRY: 2,
+    TAG_IMAGE_LATEST: 30,
     USE_K8S_PROXY: ""
     }
 
 def log_debug(m):
     msg = "{}: {}".format(inspect.stack()[1][3], m)
-    print(msg)
```
Comment (renukamanavalan) [inline src/sonic-ctrmgrd/ctrmgr/ctrmgrd.py:102]: remove it

**Review by renukamanavalan — COMMENTED** [review]
Looks good to me. 
Just take care of "any possible loop hole" before tagging as local.
If you are doing this, we could remove the "tagging as local" from the reboot script.

File: src/sonic-ctrmgrd/ctrmgr/ctrmgrd.py (hunk 1 above)
Comment (lixiaoyuner) [reply src/sonic-ctrmgrd/ctrmgr/ctrmgrd.py:102]: Okay.

File: src/sonic-ctrmgrd/ctrmgr/ctrmgrd.py [hunk 2]
```diff
@@ -426,6 +430,40 @@ def do_join(self, ip, port, insecure):
                     format(remote_ctr_config[JOIN_RETRY], self.start_time))
 
 
+def tag_latest_image(server, feat, docker_id, image_ver):
+    res = 1
+    if not UNIT_TESTING:
+        status = os.system("docker ps |grep {} >/dev/null".format(docker_id))
```
[the remark this answers is not in the document]
Comment (lixiaoyuner) [reply src/sonic-ctrmgrd/ctrmgr/ctrmgrd.py:436]: Currently, I set "tag_latest_image_on_wait_seconds" = 600s, it means when a k8s pod running for over 10 minutes(by monitor the pod's container id), I will tag the container's image. About this default value, do you have any ideas?

**Comment by lixiaoyuner** [comment]
> Looks good to me. Just take care of "any possible loop hole" before tagging as local. If you are doing this, we could remove the "tagging as local" from the reboot script.

Thanks @renukamanavalan, about possible loop hole, let me try to test this part of code carefully. And about removing "tagging as local" in reboot script, I don't think there will be some conflicts with PR, maybe remove it int the future if needed?

**Comment by renukamanavalan** [comment]
> > Looks good to me. Just take care of "any possible loop hole" before tagging as local. If you are doing this, we could remove the "tagging as local" from the reboot script.
> 
> Thanks @renukamanavalan, about possible loop hole, let me try to test this part of code carefully. And about removing "tagging as local" in reboot script, I don't think there will be some conflicts with PR, maybe remove it int the future if needed?

We will do it as one set. It is most likely to slip through the cracks otherwise. We just want one way of tagging local.
Just one more: Make sure, after tagged as local, the next service restart will let systemd take over and we don't need kube to manage this anymore.

File: src/sonic-ctrmgrd/ctrmgr/ctrmgrd.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (renukamanavalan) [reply src/sonic-ctrmgrd/ctrmgr/ctrmgrd.py:436]: May be not. Suggest
1) Try restarting service during this period
2) Kill the main process in the container that will make the container exit within this period
3) Any more .... ;)

[the remark this answers is not in the document]
Comment (lixiaoyuner) [reply src/sonic-ctrmgrd/ctrmgr/ctrmgrd.py:436]: The situation you say should be:
- k8s schedules container and remote_state is running
- ctrmgrd capture remote_state ready -> running during last process, register timer_handler(600s, tag_function(previous_docker_id))
- during the 600s, systemctl restart feature
The current reactions are:
- feature systemd service will stop -> start -> wait, when stop, set remote_state=stopped, when start, not docker start, when wait, wait docker_id
- k8s finds its scheduled container is killed, try to reschedule, container_up set remote_state=pending
- ctrmgrd captures remote_state=pending, set remote_state=ready
- container_up set remote_state=running and set docker_id
- ctrmgrd capture remote_state ready -> running, will register timer_handler(600s, tag_function(current_docker_id))
- and tag_latest_image_on_wait_seconds is up, tag function finds the previous_docker_id is disappeared, will do nothing
- and tag_latest_image_on_wait_seconds is up, tag function finds the current_docker_id, tag latest.
Have tested this case and some other cases, all good.
Any other corner cases you want me to test?

And by the way, about removing 'tag as local' from reboot script, you want me to do it in the PR or in the future? I want to do in the future, because it seems redundant, but no real impacts. And it can ensure tag as local when reboot, maybe necessary for us.

[the remark this answers is not in the document]
Comment (renukamanavalan) [reply src/sonic-ctrmgrd/ctrmgr/ctrmgrd.py:436]: I prefer we do it in this PR, as this PR's major purpose is to tag local.

One more Q: After you tag as "local" when you do service restart, who wins, systemd or kube. 

I prefer systemd takes over. This would add the new version in label that should auto-block kube from deploying. 
Can you check this out too?

[the remark this answers is not in the document]
Comment (lixiaoyuner) [reply src/sonic-ctrmgrd/ctrmgr/ctrmgrd.py:436]: Tag as local is in reboot script in sonic-utilities repo, removing from reboot script PR is here. [#](https://github.com/sonic-net/sonic-utilities/pull/2451)

Q: After you tag as "local" when do service restart, who wins?
    Kube wins.
    I see label such as feature_version_enabled=false will be set only when containers start up as local. Actually, when we restart the service, systemd set remote_state=stopped so that systemd will wait k8s to reschedule. K8s finds its container is down, will reschedule immediately. In this process, container has no change to start up as local to set feature_version_enabled=false. So, there is no auto-block.
About this problem, I think it's right that service is taken over by kube in this situation, because the set-owner is kube and k8s is connected. Taken over by kube should be the proper action.


By the way. I want to remove the instance_higher checker, because we want to support fallback in the future, if post-check failed, we could fall back to the previous version. What's your concern about this before, is there any necessary things so that we always need the higher version?

[the remark this answers is not in the document]
Comment (renukamanavalan) [reply src/sonic-ctrmgrd/ctrmgr/ctrmgrd.py:436]: Two points:
1) Please merge the reboot fix PR into this PR -- Like to keep all related in one PR. 
    For example, if we revert one, it should affect both. it should both either stay or go as one piece.
2) I agree with your explanation. Earlier tagging upon reboot only did not need any change, as state-db starts empty. Now that we are preponing the tagging, we need to do required updates that systemd takes over, if kube image is tagged as latest. In fact, if it makes simple, we can change from "kube running" to  "kube-transient" once you tag it latest. On next systemd restart, seeing it kube-transient" we can go local path.

No need for kube to manage after local tagging. It is more stable if systemd takes over.

[the remark this answers is not in the document]
Comment (lixiaoyuner) [reply src/sonic-ctrmgrd/ctrmgr/ctrmgrd.py:436]: Point 1: About the two PRs, the reason is that the reboot script is in another repo called sonic-utilities, seems we can't raise a PR including two repos code.

Point 2: About after tagging, it's easy to let feature go back to local after restart, but one of our visions is that k8s can collect the containers real-time metrics, if features are taken over by systemd after tagging, we can't keep collecting the metrics.

[the remark this answers is not in the document]
Comment (renukamanavalan) [reply src/sonic-ctrmgrd/ctrmgr/ctrmgrd.py:436]: Point 1 -- Agree.
Point 2 -- Debatable. What happens if device gets rebooted ? We have STATE-DB for counters/metrics.

Let it be so for now.

**Review by renukamanavalan — APPROVED** [review]

File: src/sonic-ctrmgrd/ctrmgr/container_startup.py [hunk 3]
```diff
@@ -229,14 +229,14 @@ def container_up(feature, owner, version):
...
 
-        if not instance_higher(feature, state_data[VERSION], version):
-            # TODO: May Remove label <feature_name>_<version>_enabled
-            # Else kubelet will continue to re-deploy every 5 mins, until
-            # master removes the lable to un-deploy.
-            #
-            do_freeze(feature, "bail out as current deploy version {} is not higher".
-                    format(version))
-            return
+        # if not instance_higher(feature, state_data[VERSION], version):
```
Comment (qiluo-msft) [inline src/sonic-ctrmgrd/ctrmgr/container_startup.py:232]: ># if not instance_higher(feature, state_data[VERSION], version): [](http://example.com/codeflow?start=8&length=64)

Do not just comment code, remove them if they are dead.

File: src/sonic-ctrmgrd/ctrmgr/ctrmgrd.py [hunk 4]
```diff
@@ -88,18 +88,20 @@
...
     JOIN_RETRY: 10,
     LABEL_RETRY: 2,
+    TAG_IMAGE_LATEST: 30,
     USE_K8S_PROXY: ""
     }
 
 def log_debug(m):
     msg = "{}: {}".format(inspect.stack()[1][3], m)
-    print(msg)
+    #print(msg)
```
Comment (qiluo-msft) [inline src/sonic-ctrmgrd/ctrmgr/ctrmgrd.py:104]: >#print(msg) [](http://example.com/codeflow?start=4&length=11)

Remove the debug code.

File: src/sonic-ctrmgrd/ctrmgr/ctrmgrd.py [hunk 5]
```diff
@@ -172,11 +176,11 @@ def register_db(self, db_name):
             self.db_connectors[db_name] = swsscommon.DBConnector(db_name, 0)
 
 
-    def register_timer(self, ts, handler):
+    def register_timer(self, ts, handler, args=()):
```
Comment (qiluo-msft) [inline src/sonic-ctrmgrd/ctrmgr/ctrmgrd.py:179]: >args=() [](http://example.com/codeflow?start=42&length=7)

Even it is allowed to use empty tuple (which is immutable) as default value, high suggest using `None` as default value.

File: src/sonic-ctrmgrd/ctrmgr/kube_commands.py [hunk 6]
```diff
@@ -332,12 +332,12 @@ def _do_reset(pending_join = False):
...
 def _do_join(server, port, insecure):
-    KUBEADM_JOIN_CMD = "kubeadm join --discovery-file {} --node-name {} --apiserver-advertise-address {}"
+    KUBEADM_JOIN_CMD = "kubeadm join --discovery-file {} --node-name {}"
     err = ""
     out = ""
     ret = 0
     try:
-        local_ipv6 = _get_local_ipv6()
+        #local_ipv6 = _get_local_ipv6()
         #_download_file(server, port, insecure)
```
Comment (qiluo-msft) [inline src/sonic-ctrmgrd/ctrmgr/kube_commands.py:341]: Delete dead code.
